### DISCUSSION TITLE
What is keeping you from using this component?

### DISCUSSION THREAD

**Discussion by KircMax** [siemens/simatic-s7-webserver-api#19] [General]
Hello everyone,

So I would like to get to know pain points some of you might have or generally what is keeping you from using the library.
I'd assume some reasons might be:

1. The Readme doesn't really tell me the benefits
2. I don't understand how to put this into a working environment
3. It doesn't fit for my use case (because...)

Please, feel free to add anything and I'd love to hear some opinions on what you might need.

**Comment by KircMax** [comment]
For those who'd like to have a bit more explanations on how to set things up and use this component:
Feel free to check out this [Youtube Channel](https://www.youtube.com/channel/UCWZCpWmXoKUeAITMsvm3Xkg). (Feel free to ask for content!)

**Reply by Siemensplcuser** [reply]
Hi KircMax,

I am Anand. I am a PLC programmer by profession and I started learning C# and Python to understand how PLC + IT Integration works. 
I cannot see any videos in your YouTube channel. Could you please share the video link?

Thanks in advance,
Anand

**Reply by KircMax** [reply]
Hi Anand,

Sorry this kind of got lost...
You can view them on my [Youtube Channel](https://www.youtube.com/channel/UCWZCpWmXoKUeAITMsvm3Xkg)...
I had put them offline since I didnt feel like there's much demand for them but since you asked - I put them back online.

Best regards
Max

**Reply by Siemensplcuser** [reply]
Hi Max,

Thank you so much for the videos. 

Best regards,
Anand

**Comment by Mytronus** [comment]
Hey, 
first of all, nice job with this project.
and now my question.
Is there an easy way to get a list of strings of each variable in a datablock down to the last tag (Integer, reals etc)?
My goal is to save the content of a datablock in a csv-file.

**Reply by KircMax** [reply]
Hey @Mytronus,
Thank you 😄,
I think I implemented something like a BrowseRecursive so basically the PlcProgram.Browse will do the trick for you but you will probably want to browse structures inside the db aswell ... I think I have that in place using a recursive Browse in the ApiPlcProgramHandler ... I'm currently just on my phone, I'd have to check on my PC - if you can't find it easily feel free to ask again

**Reply by KircMax** [reply]
Try using the 
PlcProgramBrowseSetChildrenAndParentsAsync
Of the PlcProgramHandler.

**Reply by KircMax** [reply]
Hmm, does not seem so recursive after all, I might not have that recursive Method I thought I had -
You can take a look at:
PlcProgramReadStructByChildValuesAsync
But also the Browse and set Children should work for your needs when you extend your comfort method to get all sub children for the ones you found underneath the strict, etc.
I remember I had some problems with arrays of Structs inside a struct etc. - it sure needs some improvement there I guess 😃

**Reply by PTKu** [reply]
@Mytronus 
we have developed a library to browse DB recusively. This was just POC to adapt TIA project to our existing communication library infrastructure, but you can find the principle of browsing DBs: 

https://github.com/ix-ax/axsharp/blob/dev/src/AXSharp.connectors/src/AXSharp.TIA.Connector/TIA2AXSharpAdapter.cs

**Comment by PTKu** [comment]
On holiday, so just briefly. I am using the library extensively (https://github.com/ix-ax/axsharp). 
Wider adoption has more to do with the overall culture in industrial automation than anything else; the vast majority of automation engineers are clueless about the technology in use here (REST/WebAPI...) and they lack the expertise to use it. 

The library is well-documented and tested, and it just works.

Keep up the good work @KircMax we depend on you!

**Reply by KircMax** [reply]
Thank you so much @PTKu !
I'm trying my best.

**Comment by Dragontorc** [comment]
First, thanks for the project. I like the web server as we use a lot of small hedless S7-1200 G2 PLCs (with majority having IP addresses set on the device rather than in the project) and idea of making basic setup or reconfiguring handfull of parameters using web page rather than using TIA Portal is grand.
However, in case of gen. 1 PLCs, we were able to embed our custom web pages within TIA Portal project and download them together with the project. Also, we could backup the PLC using Automation Tool and restore such project to another PLC easily.
This new approach seems to relay on using WebServerManager to make download web pages separately. The issue is when someone needs to replace their PLC. Normally they could restore the backup using Automation Tool (or even built in PLC web server service page) and that would restore PLC program, hardware config and web page as well. This makes whole download thing bit painful as it requires use of two tools (TIA and WebServerManager).
Any chance to have it integrated somehow with TIA Portal (perhaps via Openess)? After all TIA means Totally Integrated Automation :-)

**Reply by KircMax** [reply]
<img width="593" height="236" alt="image" src="https://github.com/user-attachments/assets/affe1a72-dfd9-4668-b195-1ce21d87acac" />
@Dragontorc 
Are you searching for this kind of online backup using TIA Portal maybe?
go online on the device, then click "online" at the top

**Reply by Dragontorc** [reply]
Yes. This is the same backup as done with Automation Tool or using built-in PLC service webpages. 
Question: does this backup include custom web apps downloaded to PLC?  If I restore the backup to another PLC will it include custom webapps?

**Reply by KircMax** [reply]
Yes, those custom web applications are also part of the backup :)

**Reply by Dragontorc** [reply]
That’s great!
